### ISSUE TITLE
GitLeaks throwing "too many open files" error when running with the --no-git option

### ISSUE THREAD

**Issue by srkiNZ84** [zricethezav/gitleaks#594]
**Describe the bug**
When running the "gitleaks" command against a git repository that's locally checked out on Mac OS X, it throws a "too many open files" error:

```
gitleaks --path=. -v --no-git
ERRO[0000] open .git/objects/06/c465fef0df46ca67970792e4a3b31695bd6498: too many open files
INFO[0000] scan time: 485 milliseconds 934 microseconds
ERRO[0000] open .git/objects/06/c465fef0df46ca67970792e4a3b31695bd6498: too many open files
```
The reason for running with the "--no-git" option is that I would like to only scan the current revision of the repository (all the files, not just the ones referenced by the latest commit).

The error seems to be being thrown due to GitLeaks scanning the ".git" repo in particular, which has a lot of files.

Other things I've tried to work around the above with no success are:

* Running with "threads=1" to try and reduce the number of files that the program opens "gitleaks --path=. -v --threads=1 --no-git"
* Moving the ".git" directory out of the checked out repository temporarily

**To Reproduce**
On a machine running OS X 11.4, clone a large-ish Git repository. 

Install gitleaks, change directory into the repository and run:

```
gitleaks --path=. -v --no-git
```

**Expected behavior**
I would expect that gitleaks scans the directory without throwing an error and identifies any secrets in files as per normal.

**Screenshots**
N/A

**Basic Info (please complete the following information):**
 - OS: Mac OS X 11.4
 - Gitleaks Version: 7.5.0

**Additional context**
The OS X system max open file limits are set to the following (after increasing):
```
$ launchctl limit maxfiles
	maxfiles    65536          1048576
```

cc @zricethezav

**Comment by w0rmr1d3r** [comment]
Could this PR -> https://github.com/zricethezav/gitleaks/pull/590 fix this issue?

**Comment by zricethezav** [comment]
@srkiNZ84 
> The error seems to be being thrown due to GitLeaks scanning the ".git" repo in particular, which has a lot of files.

Damn, true. Thanks for pointing that out. Gotta fix that

**Comment by srkiNZ84** [comment]
> Damn, true. Thanks for pointing that out. Gotta fix that

Just note that I did try moving the ".git" directory and it's files out of the way and it was still throwing the error, though it did manage to get a bit further in terms of scanning. So, there might still be an issue even without scanning the .git repository.

**Comment by 5cr1pt** [comment]
Temporarily set `ulimit` to a larger value in linux, such as:
```bash
$ ulimit -n 524288
```
This worked for me.

**Comment by srkiNZ84** [comment]
> Temporarily set ulimit to a larger value in linux...

Yeah, I tried this as well. Not sure why it doesn't work on OS X 🤷

**Comment by jboler** [comment]
I'm seeing something similar. I've tried various different regexes to exclude all `node_modules` directory and the `.git` directories but I can't get anything to work. Do you have any suggestions of what regex to use here?

MacOS 11.4
gitleaks 7.5.0

```
$ gitleaks -v --config-path=gitleaks.toml --no-git --path=./
ERRO[0000] open node_modules/.cache/babel-loader/6d8063beb61e7aeac6d81316036a7468.json: too many open files
INFO[0000] scan time: 59 milliseconds 48 microseconds
ERRO[0000] open node_modules/.cache/babel-loader/6d8063beb61e7aeac6d81316036a7468.json: too many open files
```

gitleaks.toml
```
title = "gitleaks config"

[[rules]]
description = "JSON Web Token"
regex = '''eyJ[A-Za-z0-9-_=]+\.[A-Za-z0-9-_=]+\.?[A-Za-z0-9-_.+/=]*?'''
tags = ["key", "jwt"]

[allowlist]
description = "Allowlisted files"
files = [
  '''.*node_modules.*''',
  '''\.git''',
]
```

**Comment by zricethezav** [comment]
> Temporarily set `ulimit` to a larger value in linux, such as:
> 
> ```shell
> $ ulimit -n 524288
> ```
> 
> This worked for me.

If the code is working the way I intended it (limited number of concurrent files open), then you really shouldn't have to set `ulimit`. Something fishy is going on. One easy solution would be to scrap multi-processing of `--no-git` altogether. This would increase the scan time significantly but would ensure no more `too many files opened` errors get thrown.

Who's got opinions... lets hear em

**Comment by sarulmurugan** [comment]
i too face the same issue. Could you please provide high priority to fix this one .

**Comment by zricethezav** [comment]
@jboler @srkiNZ84 do either of you have example repos I can try reproducing this on? The concurrency pattern found in the [`nogit` scanner](https://github.com/zricethezav/gitleaks/blob/master/scan/nogit.go#L43-L154) is pretty straight forward and based off of https://www.oreilly.com/content/run-strikingly-fast-parallel-file-searches-in-go-with-sync-errgroup/

**Comment by zricethezav** [comment]
@jboler @srkiNZ84 found the issue, turns out I wasn't limiting the number of goroutines spinning up :/ 

Gonna throw a limiter on it like the other scans

**Comment by jboler** [comment]
Thanks!

**Comment by srkiNZ84** [comment]
> found the issue, turns out I wasn't limiting the number of goroutines spinning up :/

Awesome! Thank you! Sorry I wasn't more helpful in identifying the issue

**Comment by zricethezav** [comment]
https://github.com/zricethezav/gitleaks/pull/619 going into next release

**Comment by zricethezav** [comment]
@jboler @srkiNZ84 should be fixed in 7.6.0, feel free to reopen if you find it still broken

**Comment by srkiNZ84** [comment]
> should be fixed in 7.6.0, feel free to reopen if you find it still broken

have upgraded, ran the command with "--no-git" that was consistently failing before and it is working now! Thank you so much! 🙂
